pyccel/parser/semantic.py

```diff
@@ -3785,212 +3911,263 @@ def _visit_FunctionDef(self, expr):
         for t,v in templates.items():
             templates[t] = UnionTypeAnnotation(*[self._visit(vi) for vi in v])
 
+        def unpack(ann):
+            if isinstance(ann, UnionTypeAnnotation):
+                return ann.type_list
+            else:
+                return [ann]
+
         # Filter out unused templates
-        templatable_args = [a.annotation for a in expr.arguments if isinstance(a.annotation, (SyntacticTypeAnnotation, UnionTypeAnnotation))]
-        arg_annotations = [annot for a in templatable_args for annot in (a.type_list \
-                                        if isinstance(a, UnionTypeAnnotation) else [a]) \
-                                        if isinstance(annot, SyntacticTypeAnnotation)]
-        used_type_names = set(a.dtype for a in arg_annotations)
+        templatable_args = [unpack(a.annotation) for a in expr.arguments if isinstance(a.annotation, (SyntacticTypeAnnotation, UnionTypeAnnotation))]
+        arg_annotations  = [annot for a in templatable_args for annot in a if isinstance(annot, SyntacticTypeAnnotation)]
+        used_type_names  = set(a.dtype for a in arg_annotations)
         templates = {t: v for t,v in templates.items() if t in used_type_names}
 
+        tmp_templates = {}
+        new_expr_args = []
+        for a in expr.arguments:
+            if isinstance(a.annotation, UnionTypeAnnotation):
+                annotation = [aa.type_list if isinstance(aa, UnionTypeAnnotation) else [aa] for aa in a.annotation]
+                annotation = [aa for a in annotation for aa in a]
+            else:
+                annotation = [a.annotation]
+            if len(annotation)>1:
+                tmp_template_name = a.name + '_' + random_string(12)
+                tmp_templates[tmp_template_name] = UnionTypeAnnotation(*[self._visit(vi) for vi in annotation])
+                dtype_symb = PyccelSymbol(tmp_template_name, is_temp=True)
+                dtype_symb = SyntacticTypeAnnotation(dtype_symb)
+                var_clone = AnnotatedPyccelSymbol(a.var.name, annotation=dtype_symb)
+                new_expr_args.append(FunctionDefArgument(var_clone, bound_argument = a.bound_argument,
+                                        value=a.value, kwonly=a.is_kwonly, annotation=dtype_symb))
+            else:
+                new_expr_args.append(a)
+
+        templates.update(tmp_templates)
         template_combinations = list(product(*[v.type_list for v in templates.values()]))
         template_names = list(templates.keys())
         n_templates = len(template_combinations)
-
         # this for the case of a function without arguments => no headers
         interface_name = name
         interface_counter = 0
+        is_interface = n_templates > 1
+        annotated_args = [] # collect annotated arguments to check for argument incompatibility errors
+        for f in interface_funcs:
+            self.scope.remove_symbol(f.name)
```

Comment:
in the case of an Interface, every time we pass through this [line](https://github.com/pyccel/pyccel/blob/d9bf2a422a9d87a03759cc6504d55d84f85cebaf/pyccel/parser/semantic.py#L3912) a new incremented symbol is added even if the function is not annotated, at the end we end up with function numbers that is almost random because we pass through that line many times depending on when the user calls the function. as opposed to passing through that line one time as in the past.

Reply:
Does this mean that you are relying on the names appearing in the same order to recognise the previously annotated function? Template functions are not always printed in the same order. This is due to the use of `set` a few lines earlier

Reply:
yes, I rely on the names to be the same, as for the set it should always give the same answer for the same elements

Reply:
but probably I shouldn't rely on the names to be the same, as it can cause bugs, but still, the line probably should be kept so that we don't have random names

Reply:
Yes, but sets are unordered so while it contains the same values they are not necessarily in the same order. That means that if the names are generated in the same order then they won't necessarily refer to the same functions

Reply:
> but still, the line probably should be kept so that we don't have random names

I agree. It is this line that I am really worried about:
https://github.com/pyccel/pyccel/pull/1728/files#diff-9ba0c438140113a543a41e16000b939989e841d4fc6a6a4112b5c8e5d8807524R3986

Although if there is a way to avoid the reliance on the name here we can move the name generation to after the `continue`s which would remove the need for the calls to `remove_symbol`

Reply:
yes, I shouldn't use it, I will use another way to get the function, by comparing the arguments type

Reply:
`point` in a try/except?

Reply:
done in [3f07695](https://github.com/pyccel/pyccel/pull/1728/commits/3f07695f69a94a35bd3527899ec10cff00f61e10)